File: src/bot.js
import { ChannelType, PermissionFlagsBits } from 'discord.js';
import { loadConfig } from './utils/config.js';
import { createLogger } from './utils/logger.js';
import { ConfigError } from './utils/errors.js';
import DatabaseService from './services/database.js';
import { createWebsocketClient } from './websocket/client.js';
import registerHandlers from './handlers/index.js';
import { updateStatus } from './actions/updateStatus.js';
import { sendOneTimeButton } from './actions/sendMessage.js';

class Bot {
  constructor() {
    this.config = null;
    this.logger = null;
    this.client = null;
    this.database = null;
    this.heartbeat = null;
  }

  async start() {
    this.config = await loadConfig();
    this.logger = createLogger(this.config.logging?.level);
    this.database = new DatabaseService(this.logger, this.config.storage?.stateFile);

    const { client, heartbeat } = createWebsocketClient(this.logger, this.config);
    this.client = client;
    this.heartbeat = heartbeat;

    registerHandlers(this.client, {
      bot: this,
      logger: this.logger,
      config: this.config,
      database: this.database,
    });

    const token = this.getToken();
    if (!token) {
      throw new ConfigError('DISCORD_TOKEN is not set in environment or config');
    }

    await this.client.login(token);
  }

  getToken() {
    return process.env.DISCORD_TOKEN || this.config?.bot?.token;
  }

  getDevGuildId() {
    return process.env.DEV_GUILD_ID || this.config?.bot?.devGuildId;
  }

  async handleReady() {
    this.logger.info(`Logged in as ${this.client.user?.tag ?? 'unknown user'}`);
    this.client.guilds.cache.forEach((guild) => {
      this.logger.debug(`Guild: ${guild.name} (${guild.id})`);
    });

    await updateStatus(this.client, this.config?.bot?.presence, this.logger);
    await this.sendOneTimeMessage();
  }

  async sendOneTimeMessage() {
    const targetGuildId = this.getDevGuildId();
    if (!targetGuildId) {
      this.logger.warn('DEV_GUILD_ID not set; skipping one-time message.');
      return;
    }

    const state = await this.database.getState();
    if (state[targetGuildId]) {
      this.logger.info(`One-time button already created for guild ${targetGuildId}, skipping.`);
      return;
    }

    const guild = this.client.guilds.cache.get(targetGuildId);
    if (!guild) {
      this.logger.warn(`Bot is not a member of guild ${targetGuildId} — skipping one-time message.`);
      return;
    }

    let channel = guild.systemChannel;
    if (!channel) {
      channel = guild.channels.cache.find(
        (c) => c.type === ChannelType.GuildText
          && c.permissionsFor(this.client.user)?.has(PermissionFlagsBits.SendMessages),
      );
    }

    if (!channel) {
      this.logger.warn(`No channel found in guild ${targetGuildId} where bot can send messages.`);
      return;
    }

    try {
      await sendOneTimeButton(channel, this.logger);
      await this.database.markGuildMessageSent(targetGuildId);
    } catch (error) {
      this.logger.error(`Failed to send one-time message to guild ${targetGuildId}: ${error.message}`);
    }
  }
}

export default Bot;

File: src/bot.py
"""Main bot class."""
import os
import discord
from src.events.registry import register_events
from src.utils.logger import logger


class Bot:
    """Discord bot class."""

    def __init__(self, token: str = None, intents: discord.Intents = None):
        """
        Initialize the bot.

        Args:
            token: Discord bot token
            intents: Discord intents
        """
        if intents is None:
            intents = discord.Intents.default()
            intents.guilds = True
            intents.guild_messages = True
            intents.message_content = True

        self.client = discord.Client(intents=intents)
        self.token = token or os.getenv('DISCORD_TOKEN')
        self.is_running = False

    def setup(self):
        """Setup bot event handlers."""
        register_events(self.client)
        logger.info('Bot setup complete')

    async def start(self):
        """Start the bot."""
        if self.is_running:
            logger.warn('Bot is already running')
            return

        if not self.token:
            raise ValueError('No Discord token provided')

        self.setup()

        try:
            self.is_running = True
            await self.client.start(self.token)
            logger.info('Bot started successfully')
        except Exception as err:
            logger.error('Failed to start bot', {'error': str(err)})
            self.is_running = False
            raise

    async def stop(self):
        """Stop the bot."""
        if not self.is_running:
            logger.warn('Bot is not running')
            return

        try:
            await self.client.close()
            self.is_running = False
            logger.info('Bot stopped')
        except Exception as err:
            logger.error('Error stopping bot', {'error': str(err)})
            raise

    def get_client(self) -> discord.Client:
        """Get the Discord client."""
        return self.client
